tools/clang/unittests/HLSLExec/CoopVec.h

```diff
@@ -354,6 +359,520 @@ GetMatrixSrcDataType(D3D12_LINEAR_ALGEBRA_DATATYPE MatrixInterpretation) {
+  }
+}
+
+class TestVector {
+private:
+  size_t NumVectors = 0;
+  size_t VectorSize = 0;
+  size_t ElementSize = 0;
+  size_t Stride = 0;
+  size_t TotalBytes = 0;
+  uint8_t *Buffer = nullptr;
+
+public:
+  TestVector(size_t NumVectors, size_t VectorSize, size_t ElementSize,
+             size_t Alignment = 16)
+      : NumVectors(NumVectors), VectorSize(VectorSize),
+        ElementSize(ElementSize) {
+    if (NumVectors == 0)
+      throw std::invalid_argument("NumVectors must be greater than 0");
+    if (VectorSize == 0)
+      throw std::invalid_argument("VectorSize must be greater than 0");
+    if (ElementSize == 0)
+      throw std::invalid_argument("ElementSize must be greater than 0");
+
+    const size_t VectorBytes = VectorSize * ElementSize;
+    Stride = ((VectorBytes + Alignment - 1) / Alignment) * Alignment;
+    TotalBytes = Stride * NumVectors;
+
+    void *Ptr = nullptr;
+#ifdef _MSC_VER
+    Ptr = _aligned_malloc(TotalBytes, Alignment);
+#else
+    Ptr = std::aligned_alloc(Alignment, TotalBytes);
+#endif
+    Buffer = reinterpret_cast<uint8_t *>(Ptr);
+    std::fill(Buffer, Buffer + TotalBytes, (uint8_t)0xFF);
+  }
+
+  // Copy constructor
+  TestVector(const TestVector &other)
+      : NumVectors(other.NumVectors), VectorSize(other.VectorSize),
+        ElementSize(other.ElementSize), Stride(other.Stride),
+        TotalBytes(other.TotalBytes) {
+
+    void *Ptr = nullptr;
+#ifdef _MSC_VER
+    Ptr = _aligned_malloc(TotalBytes, 16);
+#else
+    Ptr = std::aligned_alloc(16, TotalBytes);
+#endif
+    Buffer = reinterpret_cast<uint8_t *>(Ptr);
+
+    if (other.Buffer) {
+      std::memcpy(Buffer, other.Buffer, TotalBytes);
+    }
+  }
+
+  // Move constructor
+  TestVector(TestVector &&other) noexcept
+      : NumVectors(other.NumVectors), VectorSize(other.VectorSize),
+        ElementSize(other.ElementSize), Stride(other.Stride),
+        TotalBytes(other.TotalBytes), Buffer(other.Buffer) {
+
+    // Reset the source object
+    other.NumVectors = 0;
+    other.VectorSize = 0;
+    other.ElementSize = 0;
+    other.Stride = 0;
+    other.TotalBytes = 0;
+    other.Buffer = nullptr;
+  }
+
+  ~TestVector() {
+    if (Buffer) {
+#ifdef _MSC_VER
+      _aligned_free(Buffer);
+#else
+      std::free(Buffer);
+#endif
+    }
+  }
+
+  size_t getNumVectors() const { return NumVectors; }
+  size_t getVectorSize() const { return VectorSize; }
+  size_t getElementSize() const { return ElementSize; }
+  size_t getStride() const { return Stride; }
+  size_t getTotalBytes() const { return TotalBytes; }
+  uint8_t *getBuffer() { return Buffer; }
+  const uint8_t *getBuffer() const { return Buffer; }
+
+  // Copy assignment operator
+  TestVector &operator=(const TestVector &other) {
+    if (this != &other) {
+      // Free existing buffer
+      if (Buffer) {
+#ifdef _MSC_VER
+        _aligned_free(Buffer);
+#else
+        std::free(Buffer);
+#endif
+        Buffer = nullptr;
+      }
```

Comment:
You'll be able to get rid of this code with the std::shared_ptr update, if you need to release the buffer you can just call reset on the std::shared_ptr